subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/fixture/PerformanceTestScenarioDefinition.groovy

```diff
@@ -0,0 +1,139 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.performance.fixture
+
+import com.fasterxml.jackson.annotation.JsonInclude
+import com.fasterxml.jackson.databind.ObjectMapper
+import groovy.transform.CompileStatic
+import groovy.transform.EqualsAndHashCode
+
+/**
+ * See .teamcity/performance-tests-ci.json
+ */
+@EqualsAndHashCode
+@CompileStatic
+class PerformanceTestScenarioDefinition {
+    List<PerformanceTestsBean> performanceTests = []
+
+    /**
+     * Verify the jsonFile has same data as the instance itself
+     */
+    void verify(File jsonFile) {
+        sort()
+        PerformanceTestScenarioDefinition definitionInJson = new ObjectMapper().readValue(jsonFile, PerformanceTestScenarioDefinition).sort()
+        if (this != definitionInJson) {
+            File oldFile = new File(jsonFile.getAbsolutePath().replaceAll(/\.json$/, ".1.json"))
+            File newFile = new File(jsonFile.getAbsolutePath().replaceAll(/\.json$/, ".2.json"))
+            writeTo(oldFile)
+            definitionInJson.writeTo(newFile)
+
+            int size = Math.min(this.performanceTests.size(), definitionInJson.performanceTests.size())
+            int firstDifferentElementIndex = size
+            for (int i = 0; i < size; ++i) {
+                if (this.performanceTests[i] != definitionInJson.performanceTests[i]) {
+                    firstDifferentElementIndex = i
+                    break
+                }
+            }
+            throw new IllegalStateException("""
+Scenario JSON file verification fails at index ${firstDifferentElementIndex}, please see ${oldFile.getAbsolutePath()} and ${newFile.getAbsolutePath()} for details
+Old: ${firstDifferentElementIndex < performanceTests.size() ? this.performanceTests[firstDifferentElementIndex] : ""}
+New: ${firstDifferentElementIndex < definitionInJson.performanceTests.size() ? definitionInJson.performanceTests[firstDifferentElementIndex] : ""}
+""")
+        }
+    }
+
+    @EqualsAndHashCode
+    static class PerformanceTestsBean {
+        /**
+         * testId : org.gradle.performance.regression.java.JavaUpToDatePerformanceTest.up-to-date assemble (parallel true)
+         * groups : [{"testProject":"largeJavaMultiProject","coverage":{"test":["linux","windows"]}}]
+         */
+        String testId
+        List<GroupsBean> groups
+
+        PerformanceTestsBean() {
+        }
+
+        PerformanceTestsBean(String testId, List<GroupsBean> groups) {
+            this.testId = testId
+            this.groups = groups
+        }
+
+        @Override
+        String toString() {
+            return "Test($testId)"
+        }
+
+        static class GroupsBean {
+            /**
+             * testProject : largeJavaMultiProject
+             * coverage : {"test":["linux","windows"]}
+             */
+
+            String testProject
+
+            @JsonInclude(JsonInclude.Include.NON_NULL)
```

Comment:
Let's put this on the class. I don't think we want to include any non-null fields.